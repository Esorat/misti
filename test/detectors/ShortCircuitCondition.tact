contract ShortCircuitTest {
    a: Int = 5;

    // Test 1: Unoptimized && condition, should warn to reorder
    fun testCondition1(): Bool {
        return (self.expensiveCheck() && (self.a > 10));
    }

    // Test 2: Unoptimized || condition, should warn to reorder
    fun testCondition2(): Bool {
        return (true || self.expensiveCheck());
    }

    // Test 3: Complex condition with both optimized and unoptimized parts
    fun testCondition3(): Bool {
        return ((self.a > 0) && self.expensiveCheck() && false); //Bad: Should warn for moving false left
    }

    // Test 4: Nested conditions
    fun testCondition4(): Bool {
        if ((self.expensiveCheck() && (self.a < 3)) || ((self.a > 10) && true)) {
            return true; // Bad: Should warn for moving constants to the left
        }
        return false;
    }

    // Test 5: No optimization needed, should not warn
    fun testCondition5(): Bool {
        return (self.a > 0 && self.expensiveCheck());
    }

    fun expensiveCheck(): Bool {
        return true;
    }
}
