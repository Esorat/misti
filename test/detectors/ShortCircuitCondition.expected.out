[LOW] ShortCircuitCondition: Consider moving constant condition to the left for short-circuit optimization
test/detectors/ShortCircuitCondition.tact:6:17:
  5 |     fun testCondition1(): Bool {
> 6 |         return (self.expensiveCheck() && (self.a > 10));
                      ^
  7 |     }
Help: Reorder conditions to evaluate constants first
See: https://nowarp.io/tools/misti/docs/detectors/ShortCircuitCondition

[LOW] ShortCircuitCondition: Consider moving constant condition to the left for short-circuit optimization
test/detectors/ShortCircuitCondition.tact:16:17:
  15 |     fun testCondition3(): Bool {
> 16 |         return ((self.a > 0) && self.expensiveCheck() && false); //Bad: Should warn for moving false left
                       ^
  17 |     }
Help: Reorder conditions to evaluate constants first
See: https://nowarp.io/tools/misti/docs/detectors/ShortCircuitCondition

[LOW] ShortCircuitCondition: Consider moving constant condition to the left for short-circuit optimization
test/detectors/ShortCircuitCondition.tact:21:14:
  20 |     fun testCondition4(): Bool {
> 21 |         if ((self.expensiveCheck() && (self.a < 3)) || ((self.a > 10) && true)) {
                    ^
  22 |             return true; // Bad: Should warn for moving constants to the left
Help: Reorder conditions to evaluate constants first
See: https://nowarp.io/tools/misti/docs/detectors/ShortCircuitCondition